smolyay/basis.py

```diff
@@ -0,0 +1,205 @@
+import abc
+import math
+
+import numpy
+
+class BasisFunction(abc.ABC):
+    """Abstract class for basis functions.
+    Data structure for basis functions that can be called as a function
+    and holds the points 
+    ``points`` property represents all the unique 1D points within the 
+    range [-1,1] of the basis function associated with a Smolyak index 
+    described by the class IndexGrid.
+
+    """
+
+    def __init__(self):
+        self._points = []
+
+    @abc.abstractproperty
+    def points(self):
+        """The points of the basis function assigned to Smolyak indices"""
+        pass
+
+
+    @abc.abstractmethod
+    def __call__(self,x):
+        """Compute term of basis function
+        Returns the output of the basis function with input x
+
+        Parameters
+        ----------
+        x : float
+            Input arguement.
+
+        Returns
+        -------
+        float
+            Value of basis function.
+        """
+        pass
+
+
+class ChebyshevFirstKind(BasisFunction):
+    r"""Basis function Chebyshev polynomials of the first kind
+    Chebyshev polynomials are a sequence of polynomials described by the
+    following recurrence relation:
+
+    ..math:
+    T_0(x) = 1
+        T_1(x) = x
+        T_{n+1}(x) = 2xT_n(x) - T_{n-1}(x)
+
+    Object describes the nth term in the set of polynomials as designated
+    by the property ``n``
+    ``points`` represents the extrema of the polynomial of this degree on
+    the domain [-1,1]
+
+    The extrema of the polynomial are calculated via the following
+    equation:
+
+    ..math:
+        x_{i}^* = -\cos\left((\frac{i-1}{n-1}\pi\right), i = 1,...,n
+
+    Parameters
+    ----------
+    n : int
+        Degree of the Chebyshev polynomial.
+    """
+
+    def __init__(self,n):
+        super().__init__()
+        self._n = n
```

Comment:
Since `n` is readonly and the formula for computing the points is simple, just take care of it in the constructor:
```suggestion
        self._n = n
        self._points = list(numpy.polynomial.chebyshev.chebpts2(n+1)) if n > 0 else [0]
```